Add compute shader derivatives sample for Vulkan extension

gpx1000 opened the pull request:
## Description

Introduce a sample demonstrating the `VK_KHR_compute_shader_derivatives` extension. Includes new shader, CMake, and Vulkan application files showcasing the use of derivative instructions in compute shaders with quad-based derivative groups.


## General Checklist:

Please ensure the following points are checked:

- [ ] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Code-Style)
- [ ] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Copyright-Notice-and-License-Template)
- [ ] I have commented any added functions (in line with Doxygen)
- [ ] I have commented any code that could be hard to understand
- [ ] My changes do not add any new compiler warnings
- [ ] My changes do not add any new validation layer errors or warnings
- [ ] I have used existing framework/helper functions where possible
- [ ] My changes do not add any regressions
- [ ] I have tested every sample to ensure everything runs correctly
- [ ] This PR describes the scope and expected impact of the changes I am making

 Note: The Samples CI runs a number of checks including:
 - [ ] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
 - [ ] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)

 If this PR contains framework changes:
 - [ ] I did a full batch run using the `batch` command line argument to make sure all samples still work properly

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [ ] I have tested the sample on at least one compliant Vulkan implementation
- [ ] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)
- [ ] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [ ] Any dependent assets have been merged and published in downstream modules
- [ ] For new samples, I have added a paragraph with a summary to the appropriate chapter in the readme of the folder that the sample belongs to [e.g. api samples readme](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/samples/api/README.adoc)
- [ ] For new samples, I have added a tutorial README.md file to guide users through what they need to know to implement code using this feature. For example, see [conditional_rendering](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/samples/extensions/conditional_rendering)
- [ ] For new samples, I have added a link to the [Antora navigation](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/antora/modules/ROOT/nav.adoc) so that the sample will be listed at the Vulkan documentation site

gpx1000 committed:
Add compute shader derivatives sample for Vulkan extension

Introduce a sample demonstrating the `VK_KHR_compute_shader_derivatives` extension. Includes new shader, CMake, and Vulkan application files showcasing the use of derivative instructions in compute shaders with quad-based derivative groups.

SaschaWillems commented on samples/extensions/compute_shader_derivatives/README.adoc:
```diff
@@ -0,0 +1,46 @@
+- See the License for the specific language governing permissions and
+- limitations under the License.
+-
+////
```
This is missing the "Where to get the source" Antora block (see readme.adoc for other samples or the readme from the samples template).

SaschaWillems commented on samples/extensions/compute_shader_derivatives/compute_shader_derivatives.cpp:
```diff
@@ -0,0 +1,269 @@
+					float    v   = data[idx * 4 + 0];
+					float    ddx = data[idx * 4 + 1];
+					float    ddy = data[idx * 4 + 2];
+					LOGI("compute-derivatives CPU: tid=({}, {}) v={} ddx={} ddy={}", x, y, v, ddx, ddy);
```
Would be better to display them in the app's GUI instead of simply outputting them to the terminal. Depending on how/where you run the sample you don't have direct access to the terminal.

gpx1000 committed:
- Add UI overlay to display compute shader derivative results
- Store and format compute results for GUI display
- Update README with GitHub repository link for sample

gary-sweet commented:
Again, not sure what I'm supposed to see here, but I'm getting the same weird random frames on the display as I saw with relaxed_extended_instruction and some console output:

```
[info] compute-derivatives CPU: tid=(0,0) v=0.000000 ddx=1.000000 ddy=4.000000
[info] compute-derivatives CPU: tid=(1,0) v=1.000000 ddx=1.000000 ddy=4.000000
[info] compute-derivatives CPU: tid=(2,0) v=2.000000 ddx=1.000000 ddy=4.000000
[info] compute-derivatives CPU: tid=(3,0) v=3.000000 ddx=1.000000 ddy=4.000000
[info] compute-derivatives CPU: tid=(0,1) v=4.000000 ddx=1.000000 ddy=4.000000
[info] compute-derivatives CPU: tid=(1,1) v=5.000000 ddx=1.000000 ddy=4.000000
[info] compute-derivatives CPU: tid=(2,1) v=6.000000 ddx=1.000000 ddy=4.000000
[info] compute-derivatives CPU: tid=(3,1) v=7.000000 ddx=1.000000 ddy=4.000000
[info] compute-derivatives CPU: tid=(0,2) v=8.000000 ddx=1.000000 ddy=4.000000
[info] compute-derivatives CPU: tid=(1,2) v=9.000000 ddx=1.000000 ddy=4.000000
[info] compute-derivatives CPU: tid=(2,2) v=10.000000 ddx=1.000000 ddy=4.000000
[info] compute-derivatives CPU: tid=(3,2) v=11.000000 ddx=1.000000 ddy=4.000000
[info] compute-derivatives CPU: tid=(0,3) v=12.000000 ddx=1.000000 ddy=4.000000
[info] compute-derivatives CPU: tid=(1,3) v=13.000000 ddx=1.000000 ddy=4.000000
[info] compute-derivatives CPU: tid=(2,3) v=14.000000 ddx=1.000000 ddy=4.000000
[info] compute-derivatives CPU: tid=(3,3) v=15.000000 ddx=1.000000 ddy=4.000000
```

iagoCL commented on samples/extensions/compute_shader_derivatives/compute_shader_derivatives.cpp:
```diff
@@ -0,0 +1,292 @@
+	add_instance_extension(VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME);
+	// Device extension providing the feature
+	add_device_extension(VK_KHR_COMPUTE_SHADER_DERIVATIVES_EXTENSION_NAME);
+	// Toolchains may still emit SPV_NV_compute_shader_derivatives; enable NV extension if available to satisfy validation
```
Can we double-check this? Both Slang and DXC should support `khr_compute_shader_derivatives`.
It’s probably better not to request the NV extension, since if the shader is using `SPV_NV_compute_shader_derivatives `that would cause issues if a device supports the KHR extension but not the NV version.
We could add a comment suggesting users update their toolchain if they encounter this problem.

iagoCL commented on samples/extensions/compute_shader_derivatives/compute_shader_derivatives.cpp:
```diff
@@ -0,0 +1,292 @@
+void ComputeShaderDerivatives::request_gpu_features(vkb::core::PhysicalDeviceC &gpu)
+{
+	// Require quads derivative group (the sample shader uses layout(derivative_group_quadsNV/derivative_group_quads_khr))
+	REQUEST_REQUIRED_FEATURE(gpu, VkPhysicalDeviceComputeShaderDerivativesFeaturesKHR, computeDerivativeGroupQuads);
```
Some implementations only support computeDerivativeGroupLinear. As a suggestion, could we add an example using computeDerivativeGroupLinear? 

The mapping from a thread to 2D coordinates isn’t immediately obvious when using 1D quads or computeDerivativeGroupLinear, so it would be useful to have an example
https://microsoft.github.io/DirectX-Specs/d3d/HLSL_SM_6_6_Derivatives.html

Ideally, this sample could have two shader versions—derivatives_linear.comp and derivatives_quad.comp, both producing identical output with a toggle to alternate between both versions.

Also, as mentioned earlier, I think the sample would be clearer and more useful if we added a SampleGrad to use the derivatives.

asuessenbach commented on samples/extensions/compute_shader_derivatives/compute_shader_derivatives.cpp:
```diff
@@ -0,0 +1,292 @@
+					float    ddy = data[idx * 4 + 2];
+					// Store as human-readable line for GUI and also keep LOGI for debug environments
+					char buf[160];
+					snprintf(buf, sizeof(buf), "tid=(%u,%u) v=%.6f ddx=%.6f ddy=%.6f", x, y, v, ddx, ddy);
```
Maybe use `std::format`, instead of `snprintf` into a fixed size buffer.

asuessenbach reviewed (commented):
Just one minor issue.

Besides that, to see something in the UI, it's not enough to override `on_update_ui_overlay`.
You have to call `draw_ui`, which requires some more framework boilerplate code...
Do we have something like a template with a framework-based hello_triangle, or so, that could be used as the base to generate a new sample?

SaschaWillems commented:
>Do we have something like a template with a framework-based hello_triangle, or so, that could be used as the base to generate a new sample?

Yes, we do. The default templates that are used when you create a new sample via our scripts include rendering the UI. That's what  new samples should use, as the also include other boilerplate like documentation.

asuessenbach commented:
> The default templates that are used when you create a new sample via our scripts include rendering the UI.

@gpx1000 Maybe it's worth to recreate these very minimalistic samples (#1418, #1419, #1420, #1421) on that template?

gpx1000 committed:
Enhance compute shader derivatives sample with practical visualization

Extends the compute shader derivatives sample to demonstrate a practical use case: gradient-based edge detection on a procedural radial pattern. Adds a two-stage rendering pipeline where a compute shader calculates spatial derivatives and gradient magnitude, writing to a storage image, then a graphics pipeline displays the result using a fullscreen triangle technique. Includes comprehensive documentation explaining derivative computation, rendering architecture, and the fullscreen triangle optimization. Requires shader storage image read/write features and shader draw parameters extension.

gpx1000 committed:
Merge branch 'main' into VK_KHR_compute_shader_derivatives

# Conflicts:
#	samples/extensions/README.adoc
#	samples/tooling/profiles/vulkan_profiles.hpp

gpx1000 committed:
Fix code alignment and remove trailing whitespace

Aligns variable declarations and inline comments to improve code readability. Removes trailing whitespace from empty lines.
Human: [Commits history]
Update: Support more than one joint in .arm files

Also adds names to the joints based on the Blender names.
Add support for loading .arm files from Blender/Armory for GPU skinning

# Conflicts:
#	framework/rendering/subpasses/forward_subpass.cpp
#	framework/scene_graph/components/sub_mesh.cpp
Fix compute shader derivatives sample descriptor set update timing

Moves descriptor set writes to prepare() to ensure resources are updated before first use, addressing potential rendering issues from accessing uninitialized descriptors.
Enhance compute shader derivatives sample with practical visualization

Extends the compute shader derivatives sample to demonstrate a practical use case: gradient-based edge detection on a procedural radial pattern. Adds a two-stage rendering pipeline where a compute shader calculates spatial derivatives and gradient magnitude, writing to a storage image, then a graphics pipeline displays the result using a fullscreen triangle technique. Includes comprehensive documentation explaining derivative computation, rendering architecture, and the fullscreen triangle optimization. Requires shader storage image read/write features and shader draw parameters extension.
Revert "add shader quad control sample demonstrating VK_KHR_shader_quad_control features."

This reverts commit d9524ac1e14e24d8b2aa23f345fbb2bba11163e3.
add shader quad control sample demonstrating VK_KHR_shader_quad_control features.
Add compute shader derivatives sample for Vulkan extension

Introduce a sample demonstrating the `VK_KHR_compute_shader_derivatives` extension. Includes new shader, CMake, and Vulkan application files showcasing the use of derivative instructions in compute shaders with quad-based derivative groups.
bash might not work with nmake
let's try nmake
try to diagnose what's happening.
try to diagnose what's happening.
astc can't find sys/time.h when using sccache.

[Message]

[Diff]
--- a/samples/extensions/compute_shader_derivatives/compute_shader_derivatives.cpp
+++ b/samples/extensions/compute_shader_derivatives/compute_shader_derivatives.cpp
@@ -175,9 +175,9 @@
 	// Require quads derivative group (the sample shader uses layout(derivative_group_quadsNV/derivative_group_quads_khr))
 	REQUEST_REQUIRED_FEATURE(gpu, VkPhysicalDeviceComputeShaderDerivativesFeaturesKHR, computeDerivativeGroupQuads);
 	// Users may switch to the linear mode by changing the shader qualifier
-
+
 	// Storage image read/write without format (required for storage images without explicit format qualifiers)
-	gpu.get_mutable_requested_features().shaderStorageImageReadWithoutFormat = VK_TRUE;
+	gpu.get_mutable_requested_features().shaderStorageImageReadWithoutFormat  = VK_TRUE;
 	gpu.get_mutable_requested_features().shaderStorageImageWriteWithoutFormat = VK_TRUE;
 }

@@ -335,7 +335,7 @@
 	color_blend_ci.pAttachments    = &blend_attachment;

 	// Dynamic state
-	VkDynamicState dynamic_states[] = {VK_DYNAMIC_STATE_VIEWPORT, VK_DYNAMIC_STATE_SCISSOR};
+	VkDynamicState                   dynamic_states[] = {VK_DYNAMIC_STATE_VIEWPORT, VK_DYNAMIC_STATE_SCISSOR};
 	VkPipelineDynamicStateCreateInfo dynamic_ci{VK_STRUCTURE_TYPE_PIPELINE_DYNAMIC_STATE_CREATE_INFO};
 	dynamic_ci.dynamicStateCount = 2;
 	dynamic_ci.pDynamicStates    = dynamic_states;
@@ -435,7 +435,7 @@

 	// Begin render pass to display the computed image and GUI
 	VkClearValue clear_values[2];
-	clear_values[0].color        = {{0.0f, 0.0f, 0.0f, 1.0f}};  // Clear to black (will be covered by image)
+	clear_values[0].color        = {{0.0f, 0.0f, 0.0f, 1.0f}};        // Clear to black (will be covered by image)
 	clear_values[1].depthStencil = {1.0f, 0};

 	VkRenderPassBeginInfo render_pass_begin_info    = vkb::initializers::render_pass_begin_info();
@@ -464,7 +464,7 @@
 	// Render the computed image as a fullscreen quad
 	vkCmdBindPipeline(cmd, VK_PIPELINE_BIND_POINT_GRAPHICS, graphics_pipeline);
 	vkCmdBindDescriptorSets(cmd, VK_PIPELINE_BIND_POINT_GRAPHICS, graphics_pipeline_layout, 0, 1, &graphics_descriptor_set, 0, nullptr);
-	vkCmdDraw(cmd, 3, 1, 0, 0);  // Draw fullscreen triangle (3 vertices)
+	vkCmdDraw(cmd, 3, 1, 0, 0);        // Draw fullscreen triangle (3 vertices)

 	// Draw the GUI overlay on top
 	draw_ui(cmd);
@@ -501,7 +501,7 @@
 		drawer.text("- Red/Yellow: Edges (high gradient magnitude)");
 		drawer.text("- Gradient magnitude = sqrt(dx^2 + dy^2)");
 		drawer.text("");
-
+
 		drawer.text("This demonstrates edge detection using compute shader");
 		drawer.text("derivatives, useful for LOD selection, filtering, and");
 		drawer.text("spatial analysis in compute pipelines.");

gpx1000 committed:
Update copyright years to 2026

Updates copyright year ranges across multiple files to include 2026. Also fixes a typo in one header (202 -> 2026) and aligns code formatting for improved readability.

Happy New Year!

gpx1000 committed:
missed one

gary-sweet commented:
This does now draw something for me, though I'm not currently convinced it's right (which could be an issue in our driver). I'm certainly not seeing any bright red or yellow. Can you share the result you're expecting?

asuessenbach commented on samples/extensions/compute_shader_derivatives/compute_shader_derivatives.cpp:
```diff
@@ -0,0 +1,514 @@
+
+	// Storage image read/write without format (required for storage images without explicit format qualifiers)
+	gpu.get_mutable_requested_features().shaderStorageImageReadWithoutFormat  = VK_TRUE;
+	gpu.get_mutable_requested_features().shaderStorageImageWriteWithoutFormat = VK_TRUE;
```
You should check for support of those two features, like
```
	if (gpu.get_features().shaderStorageImageReadWithoutFormat)
	{
		gpu.get_mutable_requested_features().shaderStorageImageReadWithoutFormat = VK_TRUE;
	}
	else
	{
		throw std::runtime_error("GPU does not support shaderStorageImageReadWithoutFormat feature, which is required for this sample.");
	}
	if (gpu.get_features().shaderStorageImageWriteWithoutFormat)
	{
		gpu.get_mutable_requested_features().shaderStorageImageWriteWithoutFormat = VK_TRUE;
	}
	else
	{
		throw std::runtime_error("GPU does not support shaderStorageImageWriteWithoutFormat feature, which is required for this sample.");
	}
```

gpx1000 committed:
Support both quad and linear derivative group modes in compute shader derivatives sample

Enhances the compute shader derivatives sample to support both derivative group modes (quads and linear) with runtime fallback. Splits shader into two variants (quad/linear), selects appropriate shader based on device capabilities, and prioritizes quads when available. Adds validation for required storage image features and updates documentation to clarify KHR extension requirement over deprecated NV variant.

gpx1000 committed:
Add screenshot to compute shader derivatives sample README

gpx1000 committed:
Fix code formatting and alignment in compute shader derivatives sample

Aligns variable declarations, adjusts whitespace, and removes trailing spaces to improve code readability and maintain consistent formatting throughout the sample.

gpx1000 commented:
Gary, I added a screenshot; see if that helps.

gary-sweet commented:
> Gary, I added a screenshot; see if that helps.

That's great, thanks. It matches what I see with our driver, so I'm happy with that.

SaschaWillems reviewed (dismissed):
LGTM 👍🏻

Works fine on Win 11 + NV RTX 470

gpx1000 committed:
Merge branch 'main' into VK_KHR_compute_shader_derivatives

gpx1000 committed:
Merge branch 'main' into VK_KHR_compute_shader_derivatives

# Conflicts:
#	antora/modules/ROOT/nav.adoc
#	framework/vulkan_type_mapping.h
#	samples/extensions/README.adoc
#	third_party/vulkan

gpx1000 committed:
fix build errors

gpx1000 committed:
clang-format
